### PR TITLE
update compiler warning configuration 

### PR DESCRIPTION
**Pull request by jluebbe**
Add some more useful warnings and allow declarations after statements. This makes is easier to declare (temporary) variables where they are
initialized. Together with g_auto*, this makes code easier to read.

### CODE REVIEW

File: meson.build [hunk 1]
```diff
@@ -51,16 +51,24 @@ conf.set10('ENABLE_CREATE', get_option('create'))
 conf.set10('ENABLE_JSON', jsonglibdep.found())
 
 c_warn_flags = '''
+  -Walloca
   -Warray-bounds
   -Wcast-align
   -Wcast-align=strict
+  -Wdangling-pointer
```
Comment (ejoerns) [inline meson.build:58]: Covered by `-Wall` already (meson `warning_level=2`).

**Review by ejoerns — CHANGES_REQUESTED** [review]

**Comment by jluebbe** [comment]
Instead of checking multiple compilers (and rechecking on updates), I'd tend to err on the side of having some options already covered by -Wall/-Werror in the list.

**Comment by ejoerns** [comment]
> Instead of checking multiple compilers (and rechecking on updates), I'd tend to err on the side of having some options already covered by -Wall/-Werror in the list.

Maybe it's a better option to describe than how the options to enable are selected? Otherwise, this also looks a bit 'random' and might not be easier to maintain in the end. One could manually enable all GCC `-Wall` options and have it look like a stricter checking.

I'm highly unsure about what's a good strategy if not only looking at GCC. Thus if you find the approach to enable checks based on personal heuristics the most promising, I'd tend to comply.

File: meson.build [hunk 2]
```diff
@@ -51,36 +51,37 @@ conf.set10('ENABLE_CREATE', get_option('create'))
...
-  -Wwrite-strings
-  -Wpointer-arith
-  -Wmissing-declarations
-  -Wmissing-prototypes
-  -Wstrict-prototypes
-  -Wredundant-decls
-  -Wno-unused-parameter
+  -Warray-bounds
+  -Wcast-align
+  -Wcast-align=strict
```
[the remark this answers is not in the document]
Comment (jluebbe) [reply meson.build:57]: `-Wcast-align=strict` may be unsupported even if `-Wcast-align` works.

File: meson.build [hunk 3]
```diff
@@ -51,16 +51,24 @@ conf.set10('ENABLE_CREATE', get_option('create'))
...
+  -Wdate-time
   -Wdeclaration-after-statement
   -Wformat=2
   -Wformat-nonliteral
   -Wformat-security
+  -Wformat-truncation=2
   -Wimplicit-function-declaration
   -Winit-self
   -Winline
+  -Wint-conversion
```
[the remark this answers is not in the document]
Comment (jluebbe) [reply meson.build:67]: I'm not sure and would tend to enable it explicitly.

**Pull request merged**